src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedInteractiveInitializerMethod.cs

```diff
@@ -237,7 +237,15 @@ internal TypeSymbol ResultType
             {
                 diagnostics.Add(useSiteDiagnostic, NoLocation.Singleton);
             }
-            resultType = compilation.GetTypeByReflectionType(submissionReturnType, diagnostics);
+            // if no explicit return type is set on ScriptCompilationInfo, it will default to
+            // the _host_ corlib's System.Object (typeof(object)) and should never be null.
+            // In this case, redirect the host System.Object to the compilation's target
+            // corlib's System.Object. This allows cross compiling scripts to build on one
+            // corlib and run on another.
+            // cf. https://github.com/dotnet/roslyn/issues/8506
+            resultType = submissionReturnType == typeof(object)
```

Comment:
@AlekseyTs brings up a good point. We can change ScriptCompilationInfo like so:

```C#
public Type ReturnType => ReturnTypeOpt ?? typeof(object);
internal Type ReturnTypeOpt { get; }

internal ScriptCompilationInfo(Type returnType, Type globalsType)
{
      ReturnTypeOpt = returnType;
      GlobalsType = globalsType;
}
```

and then in ```CalculateReturnType``` we can replace the first line with

```C#
var submissionReturnTypeOpt = compilation.ScriptCompilationInfo.ReturnTypeOpt;
```
and avoid equality check with typeof(object).

Note that ScriptCompilationInfo should always be non-null here, since the method is only used for compiling submissions.